submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/service/DoiService.kt

```diff
@@ -0,0 +1,125 @@
+import ebi.ac.uk.model.extensions.title
+import ebi.ac.uk.util.collections.ifNotEmpty
+import mu.KotlinLogging
+import org.springframework.core.io.FileSystemResource
+import org.springframework.http.HttpHeaders
+import org.springframework.http.MediaType.MULTIPART_FORM_DATA
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.web.reactive.function.client.WebClient
+import java.nio.file.Files
+
+private val logger = KotlinLogging.logger {}
+
+@Suppress("ThrowsCount", "ReturnCount")
+class DoiService(
+    private val webClient: WebClient,
+    private val properties: DoiProperties,
+) {
+    fun calculateDoi(accNo: String, rqt: SubmitRequest): String? {
+        val doi = rqt.submission.attributes.find { it.name == DOI.name } ?: return null
+        val previousDoi = rqt.previousVersion?.doi
+
+        if (previousDoi != null) {
+            require(doi.value == previousDoi) { throw InvalidDoiException() }
+            return previousDoi
+        }
+
+        return registerDoi(accNo, rqt)
+    }
+
+    private fun registerDoi(accNo: String, rqt: SubmitRequest): String {
+        val sub = rqt.submission
+        val title = requireNotNull(sub.title) { throw MissingTitleException() }
+        val request = DoiRequest(accNo, title, properties.uiUrl, getContributors(sub))
+        val requestFile = Files.createTempFile("${TEMP_FILE_NAME}_$accNo", ".xml").toFile()
+        FileUtils.writeContent(requestFile, request.asXmlRequest())
+
+        val headers = HttpHeaders().apply { contentType = MULTIPART_FORM_DATA }
+        val body = LinkedMultiValueMap<String, Any>().apply {
+            add(USER_PARAM, properties.user)
+            add(PASSWORD_PARAM, properties.password)
+            add(OPERATION_PARAM, OPERATION_PARAM_VALUE)
+            add(FILE_PARAM, FileSystemResource(requestFile))
+        }
+
+        webClient.post(properties.endpoint, RequestParams(headers, body))
+        logger.info { "$accNo ${rqt.owner} Registered DOI: '${request.doi}'" }
+
+        return request.doi
+    }
+
+    private fun getContributors(sub: Submission): List<Contributor> {
+        val organizations = getOrganizations(sub)
+        return sub.allSections()
+            .filter { it.type.lowercase() == AUTHOR_TYPE }
+            .ifEmpty { throw MissingDoiFieldException(AUTHOR_TYPE) }
+            .map { it.asContributor(organizations) }
+    }
+
+    private fun Section.asContributor(orgs: Map<String, String>): Contributor {
+        val attrsMap = attributes.associateBy({ it.name.lowercase() }, { it.value })
+        val names = requireNotNull(attrsMap[NAME_ATTR]) { throw InvalidAuthorNameException() }
+        val affiliation = requireNotNull(attrsMap[AFFILIATION_ATTR]) { throw MissingAuthorAffiliationException() }
+        val org = requireNotNull(orgs[affiliation]) { throw InvalidAuthorAffiliationException(names, affiliation) }
+
+        return Contributor(
+            name = names.substringBeforeLast(" ", ""),
```

Comment:
why do we need this?  if this is some kind of sanitazing. Do we need this for all attributes or is specific than DOI?

Reply:
same as bellow